cartoframes/dataset.py

```diff
@@ -24,51 +33,74 @@ class Dataset(object):
-        dataset.cc.batch_sql_client \
-               .create_and_wait_for_completion(
-                   '''BEGIN; {drop}; {create}; {cartodbfy}; COMMIT;'''
-                   .format(drop=dataset._drop_table_query(),
-                           create=dataset._create_table_from_query(query),
-                           cartodbfy=dataset._cartodbfy_query()))
-
-        return dataset
+        self.cc = context or default_context
 
-    def upload(self, with_lonlat=None, if_exists='fail'):
-        if self.df is None:
-            raise ValueError('You have to create a `Dataset` with a pandas DataFrame in order to upload it to CARTO')
+        if type == Dataset.TABLE_TYPE:
+            self.table_name = normalize_name(data)
+            if self.table_name != data:
+                warn('Table will be named `{}`'.format(data))
+        elif table_name:
+            self.table_name = normalize_name(table_name)
+            if self.table_name != table_name:
+                warn('Table will be named `{}`'.format(table_name))
 
-        if not self.exists():
-            self._create_table(with_lonlat)
-        else:
-            if if_exists == Dataset.FAIL:
-                raise NameError(('Table with name {table_name} already exists in CARTO.'
-                                 ' Please choose a different `table_name` or use'
-                                 ' if_exists="replace" to overwrite it').format(table_name=self.table_name))
-            elif if_exists == Dataset.REPLACE:
-                self._create_table(with_lonlat)
+        if type == Dataset.DATAFRAME_TYPE or type == Dataset.GEODATAFRAME_TYPE:
+            self.normalized_column_names = _normalize_column_names(data)
+
+    @classmethod
+    def from_table(cls, table_name, context=None):
+        return cls(Dataset.TABLE_TYPE, table_name, context)
+
+    @classmethod
+    def from_query(cls, query, context=None):
+        return cls(Dataset.QUERY_TYPE, query, context)
+
+    @classmethod
+    def from_dataframe(cls, dataframe):
+        return cls(Dataset.DATAFRAME_TYPE, dataframe)
 
-        self._copyfrom(with_lonlat)
+    @classmethod
+    def from_geojson(cls, geojson):
+        return cls(Dataset.GEODATAFRAME_TYPE, load_geojson(geojson))
+
+    def upload(self, with_lonlat=None, if_exists='fail'):
+        if self.type == Dataset.QUERY_TYPE and self.table_name is not None and not self.exists():
+            self.cc.batch_sql_client.create_and_wait_for_completion(
+                '''BEGIN; {drop}; {create}; {cartodbfy}; COMMIT;'''
+                .format(drop=self._drop_table_query(),
+                        create=self._create_table_from_query(self.data),
+                        cartodbfy=self._cartodbfy_query()))
+        elif self.type == Dataset.DATAFRAME_TYPE or self.type == Dataset.GEODATAFRAME_TYPE:
+            if self.data is None:
+                raise ValueError('You have to create a `Dataset` with a pandas.DataFrame to upload it to CARTO')
```

Comment:
Now that we support pandas and geopandas, maybe we could update the error message. It could be a good idea, mention the static constructors we have available.